### PR TITLE
guarantee shutdown restlight server before destorying spring beans

### PR DESCRIPTION
**Pull request by Mkabaka**

### CODE REVIEW

File: restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import esa.restlight.server.util.LoggerUtils;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.event.ContextClosedEvent;
+import org.springframework.context.support.AbstractApplicationContext;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The class is designed to guarantee the {@link RestlightServer#shutdown()} happens
```
Comment (alalag1) [inline restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java:29]: "This class", not "The class"

File: restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * before spring beans are destroyed. When restlight-server-shutdown-hook registered at
+ * {@link RestlightServer#start()} executes after the spring-context-shutdown-hook registered by
+ * {@link AbstractApplicationContext#registerShutdownHook()}, the {@link #onApplicationEvent(ContextClosedEvent)}
+ * will make sure {@link RestlightServer#shutdown()} execute before destroying spring beans
+ * during {@link AbstractApplicationContext#close()}. And as the the {@link RestlightServer#shutdown()} operation
```
Comment (alalag1) [inline restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java:34]: When restlight-server-shutdown-hook `is` registered at {@link RestlightServer#start()} `which will be executed after` the spring-context-shutdown-hook which is registered by  {@link AbstractApplicationContext#registerShutdownHook()}, the {@link #onApplicationEvent(ContextClosedEvent)} will make sure `that` {@link RestlightServer#shutdown()} `will be executed` before ....

File: restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightServerAutoConfigurer.java [hunk 3]
```diff
@@ -26,16 +26,23 @@
...
     @Primary
     @Qualifier(Constants.SERVER)
     public ServerStarter defaultServerStarter(AutoRestlightServerOptions options) {
         return new ServerStarter(options);
     }
+
+    @Bean
+    public RestlightGracefulShutdown restlightGracefulShutdown(List<RestlightServer> servers) {
+        return new RestlightGracefulShutdown(servers);
+    }
```
Comment (alalag1) [inline restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightServerAutoConfigurer.java:47]: do it in `ServerStarter` plz

File: restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java (hunk 1 above)
Comment (Mkabaka) [reply restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightGracefulShutdown.java]: thx, have done!

File: restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightServerAutoConfigurer.java (hunk 3 above)
Comment (Mkabaka) [reply restlight-starter/src/main/java/esa/restlight/starter/autoconfigure/RestlightServerAutoConfigurer.java]: thx, have done!

**Review by alalag1 — APPROVED** [review]

**Pull request merged**